tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -606,53 +641,87 @@ func (a *AwsProvider) createWindowsWorkerSg(infraID string, vpc *ec2.Vpc) (strin
+					portTracker[*rule.FromPort] = true
+				}
+			}
+		}
 	}
 
-	_, err = a.EC2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
-		GroupId: aws.String(sgID),
-		IpPermissions: []*ec2.IpPermission{
+	// Add ports that need to be updated in a slice ports
+	if !portTracker[WINRM_PORT] {
+		ports = append(ports, WINRM_PORT)
+	}
+	if !portTracker[sshPort] {
+		ports = append(ports, sshPort)
+	}
+	if !portTracker[rdpPort] {
+		ports = append(ports, rdpPort)
+	}
+	return ports
+}
+
+// GetRulesForSgUpdate returns a slice of rules which are required to be updated in sg with local IP
+// This serves as an input to addIngressRules function. This function returns error if failed.
+func (a *AwsProvider) GetRulesForSgUpdate(myIP string, rules []*ec2.IpPermission, vpcCidr string) ([]*ec2.IpPermission, error) {
+	var hasClusterCidrRule bool
+	rulesForUpdate := make([]*ec2.IpPermission, 0)
+
+	ports := getPortsForSg(myIP, rules)
+
+	if rules != nil {
+		// check if cluster communication rule is missing
+		for _, rule := range rules {
+			for _, ips := range rule.IpRanges {
+				if *ips.CidrIp == vpcCidr && *rule.IpProtocol == "-1" {
+					hasClusterCidrRule = true
+					break
+				}
+			}
+		}
+	}
+
+	// If it is missing, add default rule with IpProtocol value "-1"
+	if !hasClusterCidrRule {
+		rulesForUpdate = append(rulesForUpdate,
 			(&ec2.IpPermission{}).
 				// -1 is to allow all ports.
 				SetIpProtocol("-1").
 				SetIpRanges([]*ec2.IpRange{
 					(&ec2.IpRange{}).
-						SetCidrIp(*vpc.CidrBlock),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				SetFromPort(3389).
-				SetToPort(3389).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				// winrm ansible https port
-				SetFromPort(WINRM_PORT).
-				SetToPort(WINRM_PORT).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				// SSH port to be opened for
-				// TODO: add validation to check if the port is already opened, inform the user
-				// of the same and this should be a no-op
-				SetFromPort(sshPort).
-				SetToPort(sshPort).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-		},
+						SetCidrIp(vpcCidr),
+				}))
+	}
+
+	// create a list of rules to be updated in sg
+	for _, port := range ports {
+		rulesForUpdate = append(rulesForUpdate, (&ec2.IpPermission{}).
+			SetIpProtocol("tcp").
+			SetFromPort(port).
+			SetToPort(port).
+			SetIpRanges([]*ec2.IpRange{
+				(&ec2.IpRange{}).
+					SetCidrIp(myIP + "/32"),
+			}))
+		log.Info(fmt.Sprintf("Added rule with port %d to the security groups of your local IP \n", port))
+	}
+	return rulesForUpdate, nil
+}
+
+// setSgIngress attaches inbound rules for RDP from user's external IP address and all traffic within the VPC
```

Comment:
Please update the doc. Is this function where the actual AWS API call is made to update the rule? If yes, then `addIngressRules` is not correct either.